### ISSUE TITLE
XSLT DPU should support generation of UUIDs

### ISSUE THREAD

**Issue by tomas-knap** [UnifiedViews/Plugins#128]
For the SNK pipeline (COMSODE-867) we need to be able to generate UUIDs because the data lacks natural identifiers for some entities.

In order to generate UUIDs we have originally used pure XSLT timestamp-based UUID generation, but this tends to produce non-unique UUIDs; especially for larger datasets and multithreaded execution (UUIDs start to repeat after tens of thousands invocations).

An alternative is to use direct call to the java.util.UUID#randomUUID method, which produces unique UUIDs. However, direct calls to Java methods from XSLT are only allowed in commercial versions of Saxon (i.e. Saxon-PE or Saxon-EE), whereas UnifiedViews uses Saxon-HE, in which direct Java method calls are not allowed. A solution is to create a Saxon extension (extending net.sf.saxon.lib.ExtensionFunctionDefinition) that wraps the java.util.UUID#randomUUID and exposes it to XSLT, which is allowed even for Saxon-HE.

An old pull request for original UnifiedViews DPUs implementing such extension is available here. However, due to the changes in UnifiedViews it is unlikely that it can be merged in. Instead, it can be used as an inspiration for a new implementation.

**Comment by tomas-knap** [comment]
Depends on #100, which has to be solved before.

**Comment by skodape** [comment]
How is this UUIDs used? Is it possible to generate it as a parameter for template? 
- Could we have other DPU that generates this>

**Comment by jindrichmynarz** [comment]
UUIDs are used as unique keys for minting URIs for resources for which we don't have any natural key. For example, see [this function](https://github.com/jindrichmynarz/UNIMARC-to-RDF/blob/master/lib/functions.xsl#L27-33) and its [example use](https://github.com/jindrichmynarz/UNIMARC-to-RDF/blob/master/unimarc_to_rdf.xsl#L60-63).

The function generating UUIDs needs to be directly available to XSLT because the transformation may need to generate an arbitrary number of UUIDs. Therefore, it's not possible to pass a single UUID via `xsl:param`. This is why I proposed to follow the original implementation as a [Saxon extension function](https://github.com/cammeronm/DPUs/commit/cc64913a0edda0964c4da199b7f0df266335f5f6).

**Comment by tomas-knap** [comment]
Check, specify the task

**Comment by tomas-knap** [comment]
@skrchnavy Please assign, the assigned developer can ask @jindrichmynarz to clarify details

**Comment by mkl-eea-sk** [comment]
https://github.com/UnifiedViews/Plugins/tree/feature/resolve128

**Comment by tomas-knap** [comment]
@mkl-eea-sk Please do not close the issue, but just mark it as to be tested. I will then finish the workflow.

Motivation: I can check that the code is properly formatted, commented etc. Also I can verify that it is not destructive anyhow (although not probably in this case) ;-)

**Comment by jindrichmynarz** [comment]
I can't load the updated DPU into UV (ad3d699b48b8441e3144d6749e24cf08ad081aa9):

![screen shot 2015-04-08 at 14 54 26](https://cloud.githubusercontent.com/assets/198642/7045540/51aff572-ddff-11e4-9417-e0d4fee57d45.png)

Is this a problem of the DPU or do I need to update to an even newer version of UV Core?

**Comment by mkl-eea-sk** [comment]
Core pom.xml is wrong, I have reported this to Sveto.

**Comment by jindrichmynarz** [comment]
Please add documentation of the added extension function to [README.md](https://github.com/UnifiedViews/Plugins/blob/feature/resolve128/t-xslt/README.md), so that users don't need to go through the [source code](https://github.com/UnifiedViews/Plugins/blob/feature/resolve128/t-xslt/src/main/java/cz/cuni/mff/xrg/uv/transformer/xslt/UUIDGenerator.java#L24) to find out that the function is in namespace `http://linked.opendata.cz/xslt-functions` and has local name `randomUUID`; let alone finding out that the function actually exists.

**Comment by jindrichmynarz** [comment]
While after updating Core (to [this commit](https://github.com/UnifiedViews/Core/commit/3e3efd66a4d19e26a5c14b36544c33eefac9fa71)) the DPU can be loaded, it doesn't seem to work or I'm using the extension function incorrectly. Here's what I do.

I have a minimalistic XSL template that should generate an UUID for the root node of any XML document:

```xslt
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet
    xmlns:uuid="http://linked.opendata.cz/xslt-functions"
    xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    version="2.0">
    <xsl:template match="/">
        <xsl:value-of select="uuid:randomUUID()"/>
    </xsl:template>
</xsl:stylesheet>
```

The stylesheet calls the function `randomUUID` from the namespace `http://linked.opendata.cz/xslt-functions` as Saxon's [documentation suggests](http://saxon.sourceforge.net/saxon6.5/extensibility.html#Writing-extension-functions). However, when I run a pipeline with this stylesheet used in the updated t-xslt DPU, I get a compile error:

```
eu.unifiedviews.dpu.DPUException: Cannot compile XSLT template.
	at eu.unifiedviews.helpers.dpu.context.ContextUtils.dpuException(ContextUtils.java:193) ~[uv-dpu-helpers-2.1.0-SNAPSHOT.jar:na]
	at cz.cuni.mff.xrg.uv.transformer.xslt.XsltExecutor.<init>(XsltExecutor.java:136) ~[na:na]
	at cz.cuni.mff.xrg.uv.transformer.xslt.Xslt.innerExecute(Xslt.java:85) ~[na:na]
	at eu.unifiedviews.helpers.dpu.exec.AbstractDpu.execute(AbstractDpu.java:117) ~[uv-dpu-helpers-2.1.0-SNAPSHOT.jar:na]
	at cz.cuni.mff.xrg.odcs.backend.execution.dpu.DPUExecutor.executeInstance(DPUExecutor.java:232) [backend-2.1.0-SNAPSHOT.jar:na]
	at cz.cuni.mff.xrg.odcs.backend.execution.dpu.DPUExecutor.execute(DPUExecutor.java:370) [backend-2.1.0-SNAPSHOT.jar:na]
	at cz.cuni.mff.xrg.odcs.backend.execution.dpu.DPUExecutor.run(DPUExecutor.java:452) [backend-2.1.0-SNAPSHOT.jar:na]
	at java.lang.Thread.run(Thread.java:724) ~[na:1.7.0_25]
Caused by: net.sf.saxon.s9api.SaxonApiException: Stylesheet compilation failed: 1 error reported
	at net.sf.saxon.s9api.XsltCompiler.compile(XsltCompiler.java:546) ~[na:na]
	at cz.cuni.mff.xrg.uv.transformer.xslt.XsltExecutor.<init>(XsltExecutor.java:133) ~[na:na]
	... 6 common frames omitted
Caused by: net.sf.saxon.trans.XPathException: Stylesheet compilation failed: 1 error reported
	at net.sf.saxon.style.Compilation.compileSingletonPackage(Compilation.java:97) ~[na:na]
	at net.sf.saxon.s9api.XsltCompiler.compile(XsltCompiler.java:543) ~[na:na]
	... 7 common frames omitted
```

Is this an error of the DPU or am I using it incorrectly (If so, what's its correct use?)?

**Comment by skrchnavy** [comment]
@jindrichmynarz are you able to process this xslt using saxon (or another xslt processor)? Is it valid xslt file?

**Comment by jindrichmynarz** [comment]
No, it's not possible to process this XSLT by plain Saxon or another XSLT processor exactly because it requires the implemented Saxon extension.

**Comment by tomas-knap** [comment]
@mkl-eea-sk Please address the comments from Jindrich. If not clear and difficult to solve, please discuss with @jindrichmynarz how to solve this issue, so that he may use XSLT DPU with this feature on the pipelines he needs to prepare for COMSODE

**Comment by mkl-eea-sk** [comment]
I don't know how to use this feature. I have no knowledge on using XSLT with functions.

**Comment by eea04** [comment]
Gentlemen,

randomUUID works out of the box when using Saxon, no need to use custom extensions.

I.e. this transformation works (note the value of uuid namespace):

<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xmlns:uuid="java:java.util.UUID"
    exclude-result-prefixes="xs"
    version="2.0">
    
    <xsl:template match="/">
        <xsl:value-of select="uuid:randomUUID()"/>
    </xsl:template>
    
</xsl:stylesheet>

Inspired by this:
http://stackoverflow.com/questions/15365147/how-do-i-generate-a-random-uuid-in-xquery

**Comment by jindrichmynarz** [comment]
@eea04: Sure, this is mentioned in the original description of the issue. The problem with this solution is that direct calls of Java methods are allowed only in commercial editions of Saxon (Saxon-EE or Saxon-PE), while UnifiedViews uses the non-commercial Saxon-HE, where this is not allowed. This is what motivated the solution by using Saxon extension, which can be used in Saxon-HE as well.

**Comment by jindrichmynarz** [comment]
@mkl-eea-sk: I suggest to test this Saxon extension via Java code as shown in <https://github.com/twl8n/saxon-9-he-samples>. I.e. the XSL transformation is called from Java, where the extension function is registered in the XSLT processor (see [example](https://github.com/twl8n/saxon-9-he-samples/blob/master/ext_simple.java#L79)).

**Comment by skrchnavy** [comment]
fixed in d0f41cd64491c997736f8cef899cefbe6eefb49d , @jindrichmynarz to test.
there is also simplified uri name from ```http://linked.opendata.cz/xslt-functions``` to ```xslt-functions``` so you should use your template in form:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet
    xmlns:uuid="uuid-functions"
    xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    version="2.0">
    <xsl:template match="/">
        <xsl:value-of select="uuid:randomUUID()"/>
    </xsl:template>
</xsl:stylesheet>
```

**Comment by jindrichmynarz** [comment]
Using the DPU compiled from d0f41cd64491c997736f8cef899cefbe6eefb49d and UnifiedViews/Core from [this commit](https://github.com/UnifiedViews/Core/commit/791cdce39008b4553e068ba1e2f42a90b7001164), I get the following exception, when I run a pipeline with the t-xslt DPU:

```
Loading of DPURecord implementation threw the following exception: cz.cuni.mff.xrg.odcs.commons.app.module.ModuleException: org.osgi.framework.BundleException: The bundle "eu.unifiedviews.plugins.uv-t-xslt_2.1.0.SNAPSHOT [13]" could not be resolved. Reason: Missing Constraint: Import-Package: eu.unifiedviews.dataunit; version="[2.1.0,3.0.0)"
```

**Comment by jindrichmynarz** [comment]
Also the [documentation of the implemented extension](https://github.com/UnifiedViews/Plugins/issues/128#issuecomment-91314225) is still missing.

**Comment by skrchnavy** [comment]
documentation updated.
for Core you should use develop branch, commit is now  791cdce39008b4553e068ba1e2f42a90b7001164

**Comment by jindrichmynarz** [comment]
Using the DPU compiled from commit 3936165ab89ff082addea93c6bf317013665cb5a, UnifiedViews/Core@791cdce39008b4553e068ba1e2f42a90b7001164 UUID generation works for the minimal example above.

However, when I try to run the full SNK pipeline with the updated t-xslt DPU, I get `eu.unifiedviews.dpu.DPUException: Virtual path is not set for: sample.mrc` error:

```
Execution failed because: Operation failed.eu.unifiedviews.dpu.DPUException: Operation failed.
at eu.unifiedviews.helpers.dpu.extension.faulttolerance.FaultTolerance.checkFailState(FaultTolerance.java:241)
at eu.unifiedviews.helpers.dpu.extension.faulttolerance.FaultTolerance.execute(FaultTolerance.java:255)
at cz.cuni.mff.xrg.uv.transformer.xslt.Xslt.innerExecute(Xslt.java:199)
at eu.unifiedviews.helpers.dpu.exec.AbstractDpu.execute(AbstractDpu.java:117)
at cz.cuni.mff.xrg.odcs.backend.execution.dpu.DPUExecutor.executeInstance(DPUExecutor.java:232)
at cz.cuni.mff.xrg.odcs.backend.execution.dpu.DPUExecutor.execute(DPUExecutor.java:370)
at cz.cuni.mff.xrg.odcs.backend.execution.dpu.DPUExecutor.run(DPUExecutor.java:452)
at java.lang.Thread.run(Thread.java:724)
Caused by: eu.unifiedviews.dpu.DPUException: Virtual path is not set for: sample.mrc
at eu.unifiedviews.helpers.dpu.context.ContextUtils.dpuException(ContextUtils.java:205)
at cz.cuni.mff.xrg.uv.transformer.xslt.Xslt$6.action(Xslt.java:208)
at eu.unifiedviews.helpers.dpu.extension.faulttolerance.FaultTolerance.execute(FaultTolerance.java:252)
... 6 more
```

When I remove the t-xslt DPU from the SNK pipeline, I works fine. This leads me to assume that the problem with virtual path is caused by t-xslt.

@skrchnavy: Let me know if we should deal with this in a separate issue.

**Comment by skrchnavy** [comment]
@skodape to solve this isuue

**Comment by tomas-knap** [comment]
@jindrichmynarz Could you please send the whole pipeline, which causes the problems?

**Comment by tomas-knap** [comment]
It seems that VirtualPath is not properly set in the DPU before t-xslt

**Comment by jindrichmynarz** [comment]
I thought the same but then, as mentioned above, I tested the pipeline without the t-xslt and it worked fine.

**Comment by jindrichmynarz** [comment]
Exported pipeline and sample data needed to replicate the issue sent to @tomas-knap & @skodape.

**Comment by tomas-knap** [comment]
The problem is that https://github.com/OpenDataNode/UVPlugin-ISO2709ToMarcXML
does not properly sets up VirtualPath for output files. Every DPU must define VirtualPath for output file entries using VirtualPathHelper.

Solution: VirtualPath must be properly set up for ISO2709ToMarcXML

**Comment by skrchnavy** [comment]
Closed, new issue to be opened to track [this error](#issuecomment-93695772).

**Comment by jindrichmynarz** [comment]
This extension registers a namespace identified by relative URI `uuid-functions`. Relative URIs in namespace declarations [were deprecated](http://www.w3.org/TR/xml-names11/#iri-use) in 2000. For example, if a user forgets to use `exclude-result-prefixes="uuid"` (assuming `xmlns:uuid="uuid-functions"`), then RDF validation via Jena Riot parser results in a warning: 

```
WARN  {W109} The namespace URI: <uuid-functions> is relative. Such use has been deprecated by the W3C, and may result in RDF interoperability failures. Use an absolute namespace URI.
```

While this doesn't cause any immediate issue, I'd suggest to register an absolute URI as the namespace to avoid potential interoperability issues. For example, analogously to the Java package namespace of the extension (`cz.cuni.mff.xrg.uv.transformer.xslt`), we can use an absolute URI such as `http://mff.cuni.cz/xrg/uv/transformer/xslt#`.

**Comment by skrchnavy** [comment]
To avoid confusion, I think URI must be also supported and there shall be something available. Otherwise when you use such xslt file in external processor you will receive error in compilation due to non existent domain.
For this DPU relative uri works, so do we need to change it? If yes, please create bug report, DPU works as was expected.

**Comment by jindrichmynarz** [comment]
Neither XML, nor RDF URIs must be resolvable: *"XML processors treat namespace identifiers as opaque strings and never as resolvable resources"* ([source](https://msdn.microsoft.com/en-us/library/aa468565.aspx)). While linked data greatly prefers resolvable URIs, it's not a strict requirement. XLST processors shouldn't raise an error when they encounter non-resolvable namespace URI. I agree with you that the best solution would be to use a resolvable URI, but I think non-resolvable absolute URI is preferred to non-resolvable relative URI.

It's true that for the use cases so far the relative URI works fine. However, as I have mentioned above, it may raise a warning in some parsers, so I suggested this small change to avoid problems in the future.

It would be interesting to know if relative URI used as a namespace will be considered as a bug by the data quality DPUs that are to be developed.

**Comment by skrchnavy** [comment]
What about to use this namespace: xmlns:uuid="java:java.util.UUID"?
It will be compatible with prefered java solution and also using workaround in the DPU.
It have to be tested but if work it is probably the best one.

If absolute URI shall be used then we shall use something from unifiedviews.eu as this DPU is part of UV.

**Comment by jindrichmynarz** [comment]
I believe this is Saxon-specific way of doing direct Java calls. I think that shadowing the Java URI could be confusing both for XSLT processors (as you said, it needs to be tested) and people, who may expect to find [all methods of the UUID class](http://docs.oracle.com/javase/7/docs/api/java/util/UUID.html), while the extension only covers `randomUUID()`. Therefore, I think that using a different URI is a better solution.

The choice of the URI's domain depends on whether you want to take the effort to make the URI resolve to something useful. I think that using a URI in the unifiedviews.eu domain is a good idea, because it's a domain owned by the project, so it's possible to make the URIs in it resolvable.
